src/db/transaction.rs

```diff
@@ -23,49 +24,112 @@ pub struct DbTransactionPool {
     {
+        // Get connection from pool
         let db = self.pool.get().await?;
         let db2 = db.clone();
 
+        // Lock for transaction
         let result = match (self.lock_collection.clone(), self.is_read) {
             (Some(lc), true) => db.lock_for_read(lc).await,
             (Some(lc), false) => db.lock_for_write(lc).await,
             _ => Ok(()),
         };
 
+        // Handle lock error
         if let Err(e) = result {
             db.rollback().await?;
             return Err(e.into());
         }
 
-        let resp = action(db).await?;
+        // Perform the action
+        let resp = action(db).await;
 
         // XXX: lock_for_x usually begins transactions but Dbs
         // may also implicitly create them, so commit/rollback
         // are always called to finish them. They noop when no
         // implicit transaction was created (maybe rename them
         // to maybe_commit/rollback?)
+        match resp {
+            Ok(resp) => Ok((resp, db2)),
+            Err(e) => {
+                let result = db2.rollback().await;
+
+                // Handle rollback error
+                if let Err(e) = result {
+                    self.report_error(&e);
+                    Err(e.into())
+                } else {
+                    Err(e)
+                }
+            }
+        }
+    }
+
+    /// Perform an action inside of a DB transaction.
+    pub async fn transaction<'a, A, R, F>(&'a self, action: A) -> Result<R, Error>
+    where
+        A: FnOnce(Box<dyn Db<'a>>) -> F,
+        F: Future<Output = Result<R, Error>> + 'a,
+    {
+        let (resp, db) = self.transaction_internal(action).await?;
+
+        // No further processing before commit is possible
+        let result = db.commit().await;
+
+        // Handle commit error
+        if let Err(e) = result {
+            self.report_error(&e);
+            return Err(e.into());
+        }
+
+        Ok(resp)
+    }
+
+    /// Report an error to sentry, if possible
+    fn report_error(&self, error: &ApiError) {
+        // we can't queue_report here (no access to extensions)
+        // so just report it immediately with tags on hand
+        if error.is_reportable() {
+            report(
+                &self.tags,
+                sentry::integrations::failure::event_from_fail(error),
+            );
+        } else {
+            debug!("Not reporting error: {:?}", error);
+        }
+    }
+
+    /// Perform an action inside of a DB transaction. This method will rollback
+    /// if the HTTP response is an error.
+    pub async fn transaction_http<'a, A, F>(&'a self, action: A) -> Result<HttpResponse, Error>
+    where
+        A: FnOnce(Box<dyn Db<'a>>) -> F,
+        F: Future<Output = Result<HttpResponse, Error>> + 'a,
+    {
+        let (resp, db) = self.transaction_internal(action).await?;
+
+        // HttpResponse can contain an internal error
         let result = match resp.error() {
```

Comment:
Sorry, talking through a bit of code here:

🦆 I think I'm missing something here. If I understand correctly, `transaction_internal` wraps the `action`, which will return a Result. That result could be a different Result than the `transaction_internal` wrapper. That's what we're trapping for below.  (So `transaction_internal` could return an `Ok(SomeErr(), db)` correct?) 

I see we've got [a handler](https://github.com/mozilla-services/syncstorage-rs/pull/709/files#diff-fb3d577e13883be5bf8aee1c9db0e19bR64) inside of the `transaction_internal` to handle releasing the locks on error.

Reply:
So, Actix responses can contain an internal error: https://docs.rs/actix-web/2.0.0/actix_web/struct.HttpResponse.html#method.error

`transaction_internal` handles if the action returns an error, but it doesn't know if the Ok value contains an error or not (ex. HttpResponse). `transaction_http` handles `HttpResponse` explicitly and can rollback if there's an internal error in the response.

Reply:
Ah! Ok, that was the bit I was missing. Thanks!